isis/src/base/objs/XmlToJson/XmlToJson.cpp

```diff
@@ -0,0 +1,212 @@
+ * @param output A JSON object constructed from XML input.
+ * 
+ * @return ordered_json 
+ */
+ordered_json convertXmlToJson(QDomElement& element, ordered_json& output) {
+  int i = 0;
+  while (!element.isNull()) {
+    QDomElement next = element.firstChildElement();
+    if (next.isNull()){
+      ordered_json converted = convertLastChildNodeToJson(element);
+      // Simple case with no repeated tags at the same level
+      if (!output.contains(element.tagName().toStdString())){
+        output.update(converted);
+      }
+      else {
+        // There is a repeated tag at the same level in the XML, i.e: <a>val1</a><a>val2</a>. 
+        // Translated json goal: a:[val1, val2]
+        // If the converted json has an array already, append, else make it an array  
+        if (output[element.tagName().toStdString()].is_array()) {
+          output[element.tagName().toStdString()].push_back(converted[element.tagName().toStdString()]);
+        }
+        else {
+          output[element.tagName().toStdString()] = {output[element.tagName().toStdString()], converted[element.tagName().toStdString()]};
+        }
+      }
+    }
+    else {
+      // If there is already an element with this tag name at any level besides the same one,
+      // add it to a list rather than 
+      // overwriting. This is the following situation:
+      // XML: <a> <first>value1</first> </a> <a> <second>value2</second></a>
+      // JSON: a: [ {first:value1, second:value2} ]
+      if (output.contains(element.tagName().toStdString())) {
+        
+        // If it's an array already, append, else make it an array
+        if (output[element.tagName().toStdString()].is_array()) {
+          ordered_json temporaryJson;
+          convertXmlToJson(next, temporaryJson);
+          output[element.tagName().toStdString()].push_back(temporaryJson);
+        }
+        else {
+          ordered_json temporaryJson;
+          convertXmlToJson(next, temporaryJson);
+          output[element.tagName().toStdString()] = {output[element.tagName().toStdString()], temporaryJson};
+        }
+      }
+      else {
+        if (element.hasAttributes()) {
+          // If there are attributes, add them
+          ordered_json tempArea;
+          QDomNamedNodeMap attrMap = element.attributes();
+          for (int i=0; i < attrMap.size(); i++) {
+            QDomAttr attr = attrMap.item(i).toAttr(); 
+            tempArea["@"+attr.name().toStdString()] = attr.value().toStdString();
+          }
+          tempArea.update(
+              convertXmlToJson(next, output[element.tagName().toStdString()]));
+          output[element.tagName().toStdString()] = tempArea;
+        }
+        else {
+            // Otherwise, just add element and its value
+          output[element.tagName().toStdString()] = convertXmlToJson(next, output[element.tagName().toStdString()]);
+        }
+      }
+    }
+    element = element.nextSiblingElement();
+    i++;
```

Comment:
Where does this even get used? You redefine `i` in a for loop above too.

Reply:
It's not! Thanks for catching this. It must be left over from an earlier draft of this. I've pulled it out.